source/common/http/conn_manager_impl.cc

```diff
@@ -1581,6 +1598,12 @@ void ConnectionManagerImpl::ActiveStreamDecoderFilter::encodeTrailers(HeaderMapP
   parent_.encodeTrailers(nullptr, *parent_.response_trailers_);
 }
 
+void ConnectionManagerImpl::ActiveStreamDecoderFilter::encodeMetadata(
+    MetadataMapPtr&& metadata_map) {
+  parent_.response_metadata_map_ = std::move(metadata_map);
```

Comment:
I think this might work as expected given we don't have filters which can stall sending metadata, but I think if we want the codec to handle multiple metadata blocks we can't use a single pointer to hold it, or new incoming metadata will overwrite old metadata right?

I think it may work for now because we basically synchronously pass it on, but I think it might be worth sorting out how we want to deal with separate metadata objects before we add in the complexity of filters, what do you think?

Reply:
A very good point!! I struggled a lot with the possibility of multiple metadata.

I was thinking to change MetadataEncoder to hold multiple metadata. After we go through all the filters, if metadata exists, we submit it immediately like this:

ActiveStream::encodeHeaders(){
  for (; entry != encoder_filters_.end(); entry++) {
      // filter handling here
  }
  if (response_metadata_map_ != nullptr) {
     response_encoder_->encodeMetadata(*response_metadata_map_);
     response_metadata_map_.reset();
  }
  response_encoder_->encodeHeaders();
  .....
}

ActiveStream::encodeData(){
  for (; entry != encoder_filters_.end(); entry++) {
      // filter handling here
  }
  if (response_metadata_map_ != nullptr) {
     response_encoder_->encodeMetadata(*response_metadata_map_);
     response_metadata_map_.reset();
  }
  response_encoder_->encodeData();
  .....
}

The benefit is, we will submit metadata to nghttp2 before headers or data (metadata may not drained by nghttp2 immediately, but metadata will be before headers or data when it gets drained), and we may not need to worry about the end of stream since nghttp2 will make sure the end of stream comes later either in headers or in data. But this is complete hypothetical, so it may not work at all. What do you think?

Reply:
In general in Envoy filters can decide if they pass things on immediately, or pause iteration while some other asynchronous thing happens (e.g. calling out to a rate limit or DoS service to see if we're going to accept or reject request headers)

I'm trying to decide if I'm comfortable landing metadata "as-is" (sans pause functionality) given it's most often hop-by-hop, and claim that on the request path the router could buffer and send, and on the response path we can always just encode.  I feel like it's really weird to have a one-off with no pausing, but give that metadata is most often hop-by-hop maybe it's OK to land this with only one filter action (Continue) and add pause and hold-back functionality if anyone needs it.  It would certainly cut out a lot of complexity for the initial pass.

@mattklein123 do you have any thoughts here?  I know you folks are considering using this - is pure pass-through Ok for your use case as well?  Is it too weird to have one thing that passes through sans-pause or is this a reasonable thing to land on its own?

Reply:
Also if we do go this route, might as well do away with storing the metadata - the encodeMetadata could take the std::move of the metadata_map rather than doing the store/overwrite we're doing now.   But don't bother with that change until Matt weighs in :-)

Reply:
> @mattklein123 do you have any thoughts here? I know you folks are considering using this - is pure pass-through Ok for your use case as well? Is it too weird to have one thing that passes through sans-pause or is this a reasonable thing to land on its own?

For our use case I can't think of any case in which we would pause. Either way I think it's totally fine to add the filter methods in which `Continue` is the only way to proceed for now, and we can adding pausing later if needed as you say?

Reply:
Thanks for the discussion and guidelines! They are very helpful! Will launch and iterate, :D

Reply:
Cool.  I think it probably got lost in the discussion but as I think we can leave the structure as-is, let's remove metadata_map_ as a member variable.  We can have encodeMetadata take ownership of the metadata map directly - no need to store it in the parent right?

Reply:
Sorry I missed this comment earlier!! Fixed.